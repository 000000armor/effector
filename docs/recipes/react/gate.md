---
id: gate
title: Gate: a bridge between props and store
sidebar_label: Gate: a bridge between props and store
---

Imagine you have the task of transferring something from react props to the effector store.  
Suppose you pass the history object from the react-router to the store, or pass some callbacks from render-props.  
In such a situation [`Gate`](https://effector.now.sh/en/api/effector-react/gate) will help.

In this example we will write code which will not go beyond our sandbox

```js try
import {createStore, createEffect, forward} from 'effector'
import {useStore, createGate} from 'effector-react'

// Effect for api request
const fxGetTodo = createEffect({
  async handler({id}) {
    const req = await fetch(`https://jsonplaceholder.typicode.com/todos/${id}`)
    return req.json()
  },
})

// Our main store
const $todo = createStore(null).on(fxGetTodo.done, (_, {result}) => result)

const TodoGate = createGate('gate with props')

// We callfxGetTodo effect every time Gate updates its state.
forward({from: TodoGate.state, to: fxGetTodo})

TodoGate.open.watch(() => {
  //called each time when TodoGate is mounted
})
TodoGate.close.watch(() => {
  //called each time when TodoGate is unmounted
})

function Todo() {
  const todo = useStore($todo)
  const loading = useStore(fxGetTodo.pending)
  if (loading) {
    return <div>Loading...</div>
  }

  if (Object.keys(todo).length === 0) {
    return <div>empty</div>
  }
  return (
    <div>
      <p>title: {todo.title}</p>
      <p>id: {todo.id}</p>
    </div>
  )
}

// Something that's not easy to get out of react
function ComponentWithRenderProp({children}) {
  return children(React.useState(0))
}

function App() {
  return (
    <ComponentWithRenderProp>
      {([id, setId]) => (
        <>
          <button onClick={() => setId(id + 1)}>Get next Todo</button>
          {/*In this situation, we have the ability to simultaneously
          render a component and make a request, rather than wait for the component*/}
          <TodoGate id={id} />
          <Todo />
        </>
      )}
    </ComponentWithRenderProp>
  )
}

ReactDOM.render(<App />, document.getElementById('root'))
```
