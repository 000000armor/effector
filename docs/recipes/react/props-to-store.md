---
id: gate
title: Gate: props to store
sidebar_label: Example: props to store
---

Imagine you have the task of transferring something from react props to the effector store.   
Suppose you pass the history object from the react-router to the store, or pass some callbacks from render-props.  
In such a situation [`Gate`](https://effector.now.sh/en/api/effector-react/gate) will help.  

In this example we will write code which will not go beyond our sandbox
```js try
const getTodo = ({id}) =>
  fetch(`https://jsonplaceholder.typicode.com/todos/${id}`).then(response =>
    response.json()
  )

// The effect that makes some kind of api request
const fxGetTodo = createEffect().use(getTodo)

// Our main store
const $todo = createStore(null).on(fxGetTodo.done, (_, {result}) => result)

// Our gate
const TodoGate = createGate('gate with props')

// We call for an effect every time Gate updates its state.
forward({from: TodoGate.state, to: fxGetTodo})


function Todo() {
  const todo = useStore($todo)
  const loading = useStore(fxGetTodo.pending)
  if (loading) {
    return <div>Loading...</div>
  }

  if (Object.keys(todo).length === 0) {
    return <div>empty</div>
  }
  return (
    <div>
      <p>title: {todo.title}</p>
      <p>id: {todo.id}</p>
    </div>
  )
}

// Something that's not easy to get out of react
function ComponentWithRenderProp({ children }) {
  return children(React.useState(0))
}

function App() {
  return (
    <ComponentWithRenderProp>
      {([id, setId]) => (
      	<>
          <button onClick={() => setId(id + 1)}>Get next Todo</button>
          <TodoGate id={id} /> // In this situation, we have the ability to simultaneously render a component and make a request, rather than wait for the component.
          <Todo />
        </>
      )}
    </ComponentWithRenderProp>
  )
}

ReactDOM.render(<App />, document.getElementById('root'))
```
