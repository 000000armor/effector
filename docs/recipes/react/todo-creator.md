---
id: todo-creator
title: TODO creator
---

[Try it](https://share.effector.dev/CKvKkf6j)

```ts
import React from 'react'
import ReactDOM from 'react-dom'
import {createStore, createEvent, sample} from 'effector'
import {useStore, useList} from 'effector-react'

function createTodoListApi(initial: string[] = []) {
  const insert = createEvent<string>()
  const remove = createEvent<number>()
  const change = createEvent<string>()
  const reset = createEvent<void>()

  const $input = createStore<string>('')
    .on(change, (state, value) => value).reset(reset, insert)

  const $todos = createStore<string[]>(initial)
    .on(insert, (state, value) => [...state, value])
    .on(remove, (state, index) => state.filter((_, i) => i !== index))

  const submit = createEvent()
  submit.watch(e => e.preventDefault())

  sample({
    clock: submit,
    source: input,
    target: insert,
  })

  return {
    submit,
    remove,
    change: change.prepend(e => e.currentTarget.value),
    reset,
    $todos,
    $input,
  }
}

const firstTodoList = createTodoListApi(['hello, world!'])
const secondTodoList = createTodoListApi(['hello, world!'])

function TodoList({label, model}) {
  const input = useStore(model.$input)
  
  const todos = useList(model.$todos, (value, index) => (
    <li>
      {value} <button type="button" onClick={() => model.remove(index)}>Remove</button>
    </li>
  ))

  return (
    <>
      <h1>{label}</h1>
      <ul>
        {todos}
      </ul>
      <form>
        <label>Insert todo: </label>
        <input type="text" value={input} onChange={model.change}/>
        <input type="submit" onClick={model.submit} value="Insert"/>
      </form>
    </>
  )
}

function App() {
  return (
    <>
      <TodoList
        label="First todo list"
        model={firstTodoList}
      />
      <TodoList
        label="Second todo list"
        model={secondTodoList}
      />
    </>
  )
}

ReactDOM.render(<App/>, document.getElementById('root'))
```
