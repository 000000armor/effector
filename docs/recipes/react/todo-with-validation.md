---
id: todo-with-validation
title: TODO list with input validation
---

```js
import {createEvent, createStore, createEffect, restore, combine, sample} from 'effector'
import {useStore, useList} from 'effector-react'

const validate = createEffect()
const submit = createEvent()
const submitted = createEvent()
const completed = createEvent()
const changed = createEvent()
const removed = createEvent()
const $todos = createStore([])
const $todo = restore(changed, '').reset(submitted)
const $error = restore(validate.failData, '').reset(changed)

submit.watch(e => e.preventDefault())

$todos.on(submitted, (prev, next) =>
  prev.concat({text: next, completed: false})
)
$todos.on(completed, (state, index) =>
  state.map((item, i) => ({
    ...item,
    completed: index === i ? !item.completed : item.completed,
  }))
)
$todos.on(removed, (state, index) => state.filter((_, i) => i !== index))

validate.use(([todo, todos]) => {
  if (todos.some(item => item.text === todo))
    throw 'This todo is already on the list'
  if (!todo.trim().length) throw 'Required field'
  return null
})

sample({
  source: combine($todo, $todos),
  clock: submit,
  target: validate,
})

sample({
  source: $todo,
  clock: validate.done,
  target: submitted,
})

const App = () => {
  const tasks = useStore($todos)
  const todo = useStore($todo)
  const error = useStore($error)
  const list = useList($todos, (item, index) => (
    <li style={{textDecoration: item.completed ? 'line-through' : ''}}>
      <input
        type="checkbox"
        checked={item.completed}
        onChange={() => completed(index)}
      />
      {item.text}
      <button type="button" onClick={() => removed(index)} className="delete">
        x
      </button>
    </li>
  ))
  return (
    <div>
      <h1>Todos</h1>
      <form>
        <input
          className="text"
          type="text"
          name="todo"
          value={todo}
          onChange={e => changed(e.target.value)}
        />
        <button type="submit" onClick={submit} className="submit">
          Submit
        </button>
        {error && <div className="error">{error}</div>}
      </form>

      <ul style={{listStyle: 'none'}}>{list}</ul>
    </div>
  )
}

ReactDOM.render(<App />, document.getElementById('root'))
```

[Try it](https://share.effector.dev/HuK3aUI8)
