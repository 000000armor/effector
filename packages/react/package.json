{
 "name": "@effector/react",
 "version": "0.0.1",
 "main": "src/index.js",
 "license": "MIT",
 "scripts": {
  "test": "jest --config=jest.json",
  "watch": "jest --config=jest.json --watch"
 },
 "dependencies": {
  "@effector/effector": "^0.0.1",
  "invariant": "^2.2.4",
  "react": "^16.3.2"
 },
 "devDependencies": {},
 "files": ["src"]
}
