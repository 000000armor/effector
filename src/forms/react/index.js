//@flow

export {createFormField} from './createFormField'
