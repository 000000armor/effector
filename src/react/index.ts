export {createStoreConsumer} from './createStoreConsumer'
export {createComponent} from './createComponent'
export {connect} from './connect'
export {createGate, useGate} from './createGate'
export {useEvent, useList, useStore, useStoreMap, useUnit} from './nossr'
