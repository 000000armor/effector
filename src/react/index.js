//@flow

import * as React from 'react'

import type {Store} from 'effector'

export function createReactState<State>(
 store: Store<State>,
): {
 Provider: React.ComponentType<{children: React.Node}>,
 Consumer: React.ComponentType<{
  children: (current: {
   value: State,
   ref: Store<State>,
  }) => React.Node,
 }>,
} {
 const {Provider, Consumer} = React.createContext({
  value: store.getState(),
  ref: store,
 })

 class StoreProvider extends React.Component<
  {
   children: React.Node,
  },
  {
   current: {
    value: State,
    ref: Store<State>,
   },
  },
 > {
  state = {
   current: {
    value: store.getState(),
    ref: store,
   },
  }
  componentDidMount() {
   store.watch(value => this.setState({current: {value, ref: store}}))
  }
  render() {
   return <Provider value={this.state.current}>{this.props.children}</Provider>
  }
 }

 return {
  Provider: StoreProvider,
  Consumer,
 }
}

export function createStoreComponent<
 State: Object,
 Com: React.ComponentType<*>,
>(
 store: Store<State>,
 Component: Com,
): React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
 const {Provider, Consumer} = createReactState(store)
 const ConnectedComponent = props => (
  <Provider>
   <Consumer>{({value}) => <Component {...props} {...value} />}</Consumer>
  </Provider>
 )
 ConnectedComponent.displayName = `Connect(${Component.displayName
  || 'Unknown'})`
 return ConnectedComponent
}

export function connect<State: Object, Com: React.ComponentType<*>>(
 Component: Com,
): (
 store: Store<State>,
) => React.ComponentType<$Exact<$Diff<React.ElementConfig<Com>, State>>> {
 return store => createStoreComponent(store, Component)
}
