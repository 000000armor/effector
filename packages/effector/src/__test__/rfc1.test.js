//@flow
import * as React from 'react'
import TestRenderer from 'react-test-renderer'

import {
 createEvent,
 createEffect,
 createStore,
 type Event,
 type Effect,
 type Store,
} from '..'

test('createStore', () => {
 const counter = createStore(0)
 const text = createStore('')
 const store = createStore({counter, text, foo: 'bar'})
 expect(store.getState()).toMatchObject({counter: 0, text: '', foo: 'bar'})
})

test('event.to', () => {
 const counter = createStore(0)
 const text = createStore('')
 const store = createStore({counter, text, foo: 'bar'})

 const e1: Event<string, any> = createEvent('e1')
 e1.to(store, (state, payload) => ({
  ...state,
  foo: payload,
 }))

 expect(store.getState()).toMatchObject({counter: 0, text: '', foo: 'bar'})
 e1('baz')
 expect(store.getState()).toMatchObject({counter: 0, text: '', foo: 'baz'})
})

test('event lock', () => {
 const store = createStore({foo: ['_']})
 const fn1 = jest.fn()
 const fn2 = jest.fn()
 const e1: Event<string, *> = createEvent('e1')
 store.on(e1, ({foo}, payload) => ({
  foo: [...foo, payload],
 }))

 const unsub1 = e1.watch(data => {
  fn1(data)
  switch (data) {
   case 'a':
    return e1('b')
   case 'b':
    return e1('c')
  }
 })
 expect(store.getState()).toMatchObject({foo: ['_']})
 e1('a')
 expect(store.getState()).toMatchObject({foo: ['_', 'a', 'b', 'c']})

 unsub1()

 e1.watch(data => {
  fn2(data)
  switch (data) {
   case 'a':
    return e1('b')
   case 'b':
    return e1('c')
   case 'c':
    return e1('c')
  }
 })
 expect(() => {
  e1('a')
 }).toThrowErrorMatchingSnapshot()
 expect(fn1).toHaveBeenCalledTimes(3)
 expect(fn2).toHaveBeenCalledTimes(3)
 expect(store.getState()).toMatchObject({
  foo: ['_', 'a', 'b', 'c', 'a', 'b', 'c'],
 })
})

test('store.on', () => {
 const counter = createStore(0)
 const text = createStore('')
 const store = createStore({counter, text, foo: 'bar'})

 const e1 = createEvent('e1')
 store.on(e1, (state, payload) => ({
  ...state,
  foo: payload,
 }))

 expect(store.getState()).toMatchObject({counter: 0, text: '', foo: 'bar'})
 e1('baz')
 expect(store.getState()).toMatchObject({counter: 0, text: '', foo: 'baz'})
})

test('rfc1 example implementation', async() => {
 const inputText: Event<string, *> = createEvent('input text')
 const click: Event<void, *> = createEvent('click')
 const resetForm: Event<void, *> = createEvent('reset')
 const increment: Event<void, *> = createEvent('increment')

 const fetchSavedText: Effect<string, any, any, any> = createEffect(
  'fetch saved text',
 )
 fetchSavedText.use(() => Promise.resolve('~~ mock for saved text ~~'))

 const counter = createStore(0)
 const text = createStore('')
 const store = createStore({counter, text})

 counter.reset(resetForm)
 text.reset(resetForm)

 inputText
  .map(text => text.trim())
  .epic(text$ => text$.debounce(500))
  .to(text)

 increment.to(counter, state => state + 1)

 click.epic(click$ => click$.throttle(500)).watch(async() => {
  await new Promise(_ => setTimeout(_, 500))
  increment()
 })

 //  fetchSavedText.done.map(({result}) => result).to(inputText)

 const ClickedTimes = store
  .map(({counter, text}) => 'Clicked: ' + counter + ' times')
  .withProps(state => {
   console.log(state)
   expect(state).not.toBe(text)
   expect(typeof state).toBe('string')
   return <span>{state}</span>
  })

 const CurrentText = store.withProps(({counter, text}, props) => (
  <p>
   {props.prefix} {text}
  </p>
 ))

 const App = () => (
  <React.Fragment>
   <ClickedTimes />
   <CurrentText prefix="Current text: " />
  </React.Fragment>
 )
 expect(ClickedTimes).toBeDefined()
 expect(<ClickedTimes />).toBeDefined()
 expect(App).toBeDefined()
 expect(TestRenderer.create(<ClickedTimes />).toJSON()).toMatchSnapshot()
 expect(
  TestRenderer.create(<CurrentText prefix="Current text: " />).toJSON(),
 ).toMatchSnapshot()
 expect(TestRenderer.create(<App />).toJSON()).toMatchSnapshot()
})
