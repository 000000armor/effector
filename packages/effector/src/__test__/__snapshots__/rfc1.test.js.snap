// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`event lock 1`] = `"already locked"`;

exports[`rfc1 example implementation 1`] = `
<span>
  Clicked: 0 times
</span>
`;

exports[`rfc1 example implementation 2`] = `
<p>
  Current text: 
   
  
</p>
`;

exports[`rfc1 example implementation 3`] = `
Array [
  <span>
    Clicked: 0 times
  </span>,
  <p>
    Current text: 
     
    
  </p>,
]
`;

exports[`rfc1 example implementation 4`] = `
Array [
  <span>
    Clicked: 2 times
  </span>,
  <p>
    Current text: 
     
    
  </p>,
]
`;
